File: src/check.ts
import { Effect, Either, Option, Schema } from "effect";
import type { KubeTarget } from "./config.js";
import { classify, remediation, type Category } from "./classify.js";
import {
  kubectlStderr,
  kubectlStdout,
  runKubectl,
  type KubectlError as KubectlErr,
} from "./runner.js";
import { sanitizeLines } from "./sanitize.js";

export interface CheckResult {
  readonly name: string;
  readonly ok: boolean;
  readonly advisory?: boolean;
  readonly category?: Category;
  readonly message: string;
  readonly remediation?: string;
}

export interface CheckReport {
  readonly context: string;
  readonly namespace: string;
  readonly ok: boolean;
  readonly checks: ReadonlyArray<CheckResult>;
  readonly checkedAt: string;
}

export interface CheckOptions {
  readonly kubectlPath?: string;
  readonly kubeconfig?: string;
  readonly extraEnv?: NodeJS.ProcessEnv;
  readonly requestTimeoutSec?: number;
  readonly now?: () => Date;
}

interface Perm {
  readonly verb: string;
  readonly resource: string;
  readonly subresource?: string;
  readonly namespaced: boolean;
  readonly advisory: boolean;
}

const requiredPerms: ReadonlyArray<Perm> = [
  { verb: "get", resource: "sandboxes.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "list", resource: "sandboxes.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "watch", resource: "sandboxes.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "create", resource: "sandboxes.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "patch", resource: "sandboxes.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "delete", resource: "sandboxes.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "get", resource: "sandboxclaims.extensions.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "list", resource: "sandboxclaims.extensions.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "create", resource: "sandboxclaims.extensions.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "patch", resource: "sandboxclaims.extensions.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "delete", resource: "sandboxclaims.extensions.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "get", resource: "sandboxtemplates.extensions.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "list", resource: "sandboxtemplates.extensions.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "get", resource: "sandboxwarmpools.extensions.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "list", resource: "sandboxwarmpools.extensions.agents.x-k8s.io", namespaced: true, advisory: false },
  { verb: "get", resource: "pods", namespaced: true, advisory: false },
  { verb: "list", resource: "pods", namespaced: true, advisory: false },
  { verb: "create", resource: "pods", subresource: "exec", namespaced: true, advisory: false },
  { verb: "get", resource: "pods", subresource: "log", namespaced: true, advisory: false },
  { verb: "get", resource: "persistentvolumeclaims", namespaced: true, advisory: false },
  { verb: "list", resource: "persistentvolumeclaims", namespaced: true, advisory: false },
  { verb: "get", resource: "resourcequotas", namespaced: true, advisory: false },
  { verb: "list", resource: "resourcequotas", namespaced: true, advisory: false },
  { verb: "get", resource: "limitranges", namespaced: true, advisory: false },
  { verb: "get", resource: "secrets", namespaced: true, advisory: true },
  { verb: "get", resource: "storageclasses.storage.k8s.io", namespaced: false, advisory: false },
  { verb: "list", resource: "storageclasses.storage.k8s.io", namespaced: false, advisory: false },
  { verb: "get", resource: "runtimeclasses.node.k8s.io", namespaced: false, advisory: false },
  { verb: "get", resource: "namespaces", namespaced: false, advisory: false },
];

const ok = (name: string, message: string): CheckResult => ({ name, ok: true, message });

const fail = (name: string, message: string, cat: Category): CheckResult => ({
  name,
  ok: false,
  message,
  category: cat,
  remediation: remediation(cat),
});

const baseArgs = (cfg: KubeTarget, req: string, args: ReadonlyArray<string>): Array<string> => [
  "--context",
  cfg.context,
  "--request-timeout",
  req,
  ...args,
];

const namespacedArgs = (cfg: KubeTarget, req: string, args: ReadonlyArray<string>): Array<string> => [
  "--context",
  cfg.context,
  "-n",
  cfg.namespace,
  "--request-timeout",
  req,
  ...args,
];

const firstUseful = (...parts: Array<string>): string => {
  for (const p of parts) {
    const t = p.trim();
    if (t) return t;
  }
  return "unknown error";
};

const classifiedResult = (name: string, action: string, err: KubectlErr): CheckResult => {
  const stderr = kubectlStderr(err);
  const cat = classify(stderr, err._tag === "KubectlTimeoutError");
  const detail = firstUseful(stderr, kubectlStdout(err), err._tag === "KubectlSpawnError" ? err.cause : "kubectl failed");
  return fail(name, `could not ${action}: ${sanitizeLines(detail).slice(0, 500)}`, cat);
};

const parseJson = <A, I>(raw: string, schema: Schema.Schema<A, I>): A | null =>
  Option.getOrNull(Schema.decodeUnknownOption(Schema.parseJson(schema))(raw));

const permArgs = (cfg: KubeTarget, req: string, p: Perm): Array<string> => {
  const base = p.namespaced
    ? namespacedArgs(cfg, req, ["auth", "can-i", p.verb, p.resource])
    : baseArgs(cfg, req, ["auth", "can-i", p.verb, p.resource]);
  // TYPE/NAME selects an object, not a subresource.
  return p.subresource ? [...base, `--subresource=${p.subresource}`] : base;
};

const computeQuotaRe = /^(pods|cpu|memory|requests\.cpu|requests\.memory|limits\.cpu|limits\.memory|requests\.ephemeral-storage|limits\.ephemeral-storage)$/;
const storageQuotaRe = /^(persistentvolumeclaims|requests\.storage)$/;

const resourceMetadata = Schema.Struct({
  name: Schema.optional(Schema.String),
  annotations: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.String })),
});
const runtimeClass = Schema.Struct({ handler: Schema.optional(Schema.String) });
const storageClasses = Schema.Struct({
  items: Schema.optional(Schema.Array(Schema.Struct({ metadata: Schema.optional(resourceMetadata) }))),
});
const resourceQuotas = Schema.Struct({
  items: Schema.optional(Schema.Array(Schema.Struct({
    metadata: Schema.optional(resourceMetadata),
    spec: Schema.optional(Schema.Struct({
      hard: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.String })),
    })),
  }))),
});
const limitRanges = Schema.Struct({ items: Schema.optional(Schema.Array(Schema.Struct({}))) });

const shortResource = (r: string): string =>
  r
    .replaceAll(".agents.x-k8s.io", "")
    .replaceAll(".extensions", "")
    .replaceAll(".storage.k8s.io", "")
    .replaceAll(".node.k8s.io", "")
    .replaceAll("/", "-");

interface Probe {
  readonly out: string;
  readonly err: CheckResult | null;
}

type Attempt = (name: string, action: string, args: ReadonlyArray<string>) => Effect.Effect<Probe>;
type Run = (args: ReadonlyArray<string>) => Effect.Effect<string, KubectlErr>;

const checkApis = (cfg: KubeTarget, req: string, attempt: Attempt): Effect.Effect<ReadonlyArray<CheckResult>> =>
  Effect.gen(function*() {
    const checks: Array<CheckResult> = [];
    const connectivity = yield* attempt("api-connectivity", "reach the Kubernetes API with the saved context", baseArgs(cfg, req, ["api-versions"]));
    const served = connectivity.err
      ? null
      : new Set(connectivity.out.split("\n").map((line) => line.trim()).filter(Boolean));
    checks.push(connectivity.err ?? ok("api-connectivity", "API reachable with saved context"));

    for (const group of [
      {
        name: "agents-api",
        apiGroup: "agents.x-k8s.io",
        version: "v1beta1",
        want: ["sandboxes"],
        display: "Sandbox API agents.x-k8s.io/v1beta1",
      },
      {
        name: "extensions-api",
        apiGroup: "extensions.agents.x-k8s.io",
        version: "v1beta1",
        want: ["sandboxclaims", "sandboxtemplates", "sandboxwarmpools"],
        display: "Sandbox extensions API extensions.agents.x-k8s.io/v1beta1",
      },
    ]) {
      if (served !== null && !served.has(`${group.apiGroup}/${group.version}`)) {
        checks.push(fail(group.name, `${group.display} is not served`, "missing-infrastructure"));
        continue;
      }
      const result = yield* attempt(
        group.name,
        `discover ${group.display}`,
        baseArgs(cfg, req, ["api-resources", `--api-group=${group.apiGroup}`, "-o", "name"]),
      );
      if (result.err) {
        checks.push(result.err);
        continue;
      }
      const missing = group.want.filter((resource) => !result.out.toLowerCase().includes(resource));
      checks.push(missing.length > 0
        ? fail(group.name, `${group.display} served but missing: ${missing.join(", ")}`, "missing-infrastructure")
        : ok(group.name, `${group.display} served`));
    }
    return checks;
  });

const checkRuntimeClass = (cfg: KubeTarget, req: string, attempt: Attempt): Effect.Effect<CheckResult> =>
  Effect.gen(function*() {
    const result = yield* attempt("runtimeclass-gvisor", "read RuntimeClass gvisor", baseArgs(cfg, req, ["get", "runtimeclass", "gvisor", "-o", "json"]));
    if (result.err) return result.err;
    const parsed = parseJson(result.out, runtimeClass);
    if (parsed === null) return fail("runtimeclass-gvisor", "RuntimeClass gvisor returned unreadable JSON", "unknown");
    const handler = parsed.handler ?? "";
    return handler.toLowerCase().includes("runsc")
      ? ok("runtimeclass-gvisor", `RuntimeClass gvisor present (handler ${handler})`)
      : fail("runtimeclass-gvisor", `RuntimeClass gvisor handler is "${handler}", want runsc`, "missing-infrastructure");
  });

const checkStorage = (cfg: KubeTarget, req: string, attempt: Attempt): Effect.Effect<CheckResult> =>
  Effect.gen(function*() {
    const result = yield* attempt("storage", "list StorageClasses", baseArgs(cfg, req, ["get", "storageclass", "-o", "json"]));
    if (result.err) return result.err;
    const parsed = parseJson(result.out, storageClasses);
    if (parsed === null) return fail("storage", "StorageClass list returned unreadable JSON", "unknown");
    const items = parsed.items ?? [];
    if (items.length === 0) return fail("storage", "no StorageClasses available for sandbox homes", "missing-infrastructure");
    const names = items.map((item) => item.metadata?.name ?? "?");
    const defaultName = items.find((item) => item.metadata?.annotations?.["storageclass.kubernetes.io/is-default-class"] === "true")
      ?.metadata?.name;
    const defaultLabel = defaultName ? ` (default ${defaultName})` : "";
    return ok("storage", `${names.length} StorageClass(es): ${names.join(", ")}${defaultLabel}`);
  });

const checkQuota = (cfg: KubeTarget, req: string, attempt: Attempt): Effect.Effect<CheckResult> =>
  Effect.gen(function*() {
    const result = yield* attempt("budgets-quota", "read ResourceQuotas in target namespace", namespacedArgs(cfg, req, ["get", "resourcequota", "-o", "json"]));
    if (result.err) return result.err;
    const parsed = parseJson(result.out, resourceQuotas);
    if (parsed === null) return fail("budgets-quota", "ResourceQuota list returned unreadable JSON", "unknown");
    const items = parsed.items ?? [];
    if (items.length === 0) {
      return fail(
        "budgets-quota",
        "no ResourceQuota in target namespace; set explicit compute/storage budgets via the Helm chart",
        "missing-infrastructure",
      );
    }
    const names = items.map((item) => item.metadata?.name ?? "?").join(", ");
    const keys = [...new Set(items.flatMap((item) => Object.keys(item.spec?.hard ?? {})))];
    const hasCompute = keys.some((key) => computeQuotaRe.test(key));
    const hasStorage = keys.some((key) => storageQuotaRe.test(key));
    if (hasCompute && hasStorage) return ok("budgets-quota", `ResourceQuota present: ${names}`);
    const missing = [!hasCompute ? "compute (pods, cpu, memory)" : null, !hasStorage ? "storage" : null]
      .filter((name) => name !== null)
      .join(" and ");
    return fail(
      "budgets-quota",
      `ResourceQuota ${names} sets no ${missing} budget (hard: ${keys.join(", ") || "empty"})`,
      "missing-infrastructure",
    );
  });

const checkLimits = (cfg: KubeTarget, req: string, attempt: Attempt): Effect.Effect<CheckResult> =>
  Effect.gen(function*() {
    const result = yield* attempt("budgets-limits", "read LimitRanges in target namespace", namespacedArgs(cfg, req, ["get", "limitrange", "-o", "json"]));
    if (result.err) return { ...result.err, advisory: true };
    const parsed = parseJson(result.out, limitRanges);
    if (parsed === null) return { ...fail("budgets-limits", "LimitRange list returned unreadable JSON", "unknown"), advisory: true };
    const count = (parsed.items ?? []).length;
    return count === 0
      ? { ...fail("budgets-limits", "no LimitRange in target namespace", "missing-infrastructure"), advisory: true }
      : ok("budgets-limits", `${count} LimitRange(s) present`);
  });

const checkPermissions = (cfg: KubeTarget, req: string, run: Run): Effect.Effect<ReadonlyArray<CheckResult>> =>
  Effect.forEach(requiredPerms, (permission) => Effect.gen(function*() {
    const shown = permission.subresource ? `${permission.resource}/${permission.subresource}` : permission.resource;
    const name = `perm-${permission.verb}-${shortResource(shown)}`;
    const probed = yield* Effect.either(run(permArgs(cfg, req, permission)));
    const out = Either.isLeft(probed) ? kubectlStdout(probed.left) : probed.right;
    const answer = out.trim().toLowerCase();
    if (answer === "yes") return ok(name, `can ${permission.verb} ${shown}`);
    if (answer === "no" || answer.startsWith("no ")) {
      const scope = permission.namespaced ? "namespace" : "cluster scope";
      const denied = fail(name, `cannot ${permission.verb} ${shown} in ${scope}`, "denied");
      return permission.advisory ? { ...denied, advisory: true } : denied;
    }
    if (Either.isLeft(probed)) {
      const result = classifiedResult(name, `check permission ${permission.verb} ${shown}`, probed.left);
      return permission.advisory ? { ...result, advisory: true } : result;
    }
    const unexpected = fail(name, `unexpected permission answer for ${permission.verb} ${shown}: "${sanitizeLines(out)}"`, "unknown");
    return permission.advisory ? { ...unexpected, advisory: true } : unexpected;
  }), { concurrency: 6 });

export const runCheck = (cfg: KubeTarget, opts: CheckOptions): Effect.Effect<CheckReport, never> => {
  const req = `${opts.requestTimeoutSec ?? 10}s`;
  const perCallMs = Math.max((opts.requestTimeoutSec ?? 10) * 1000 + 5000, 15000);
  const run: Run = (args) => runKubectl(args, {
    kubectlPath: opts.kubectlPath,
    kubeconfig: opts.kubeconfig,
    extraEnv: opts.extraEnv,
    timeoutMs: perCallMs,
  });
  const attempt: Attempt = (name, action, args) => Effect.match(run(args), {
    onFailure: (error) => ({ out: kubectlStdout(error), err: classifiedResult(name, action, error) }),
    onSuccess: (out) => ({ out, err: null }),
  });

  return Effect.gen(function*() {
    const api = yield* checkApis(cfg, req, attempt);
    const runtime = yield* checkRuntimeClass(cfg, req, attempt);
    const storage = yield* checkStorage(cfg, req, attempt);
    const namespace = yield* attempt("namespace", `read namespace "${cfg.namespace}"`, baseArgs(cfg, req, ["get", "namespace", cfg.namespace, "-o", "json"]));
    const quota = yield* checkQuota(cfg, req, attempt);
    const limits = yield* checkLimits(cfg, req, attempt);
    const permissions = yield* checkPermissions(cfg, req, run);
    const checks = [
      ...api,
      runtime,
      storage,
      namespace.err ?? ok("namespace", `namespace "${cfg.namespace}" exists`),
      quota,
      limits,
      ...permissions,
    ];
    return {
      context: cfg.context,
      namespace: cfg.namespace,
      ok: checks.every((check) => check.ok || check.advisory),
      checks,
      checkedAt: (opts.now?.() ?? new Date()).toISOString(),
    } satisfies CheckReport;
  });
};

File: internal/kubeflock/check.go
package kubeflock

import (
	"context"
	"encoding/json/jsontext"
	"encoding/json/v2"
	"errors"
	"fmt"
	"maps"
	"slices"
	"strings"
	"time"

	"golang.org/x/sync/errgroup"
	"k8s.io/apimachinery/pkg/util/sets"
)

const (
	extensionsAPIGroup    = "extensions.agents.x-k8s.io"
	runtimeClassCheckName = "runtimeclass-gvisor"
	quotaCheckName        = "budgets-quota"
	limitsCheckName       = "budgets-limits"
)

type globalOptions struct {
	ConfigPath string
	Kubeconfig string
	Kubectl    string
	StateDir   string
}

type accessCheck struct {
	Name        string
	Verb        string
	Group       string
	Resource    string
	Subresource string
	Namespaced  bool
	Advisory    bool
}

func (c accessCheck) resourceArg() string {
	if c.Group == "" {
		return c.Resource
	}
	return c.Resource + "." + c.Group
}

func (c accessCheck) qualified() string {
	name := c.resourceArg()
	if c.Subresource != "" {
		name += "/" + c.Subresource
	}
	return name
}

var requiredAccess = []accessCheck{
	{Name: "perm-get-sandboxes", Verb: "get", Group: "agents.x-k8s.io", Resource: "sandboxes", Namespaced: true},
	{Name: "perm-patch-sandboxes", Verb: "patch", Group: "agents.x-k8s.io", Resource: "sandboxes", Namespaced: true},
	{Name: "perm-delete-sandboxes", Verb: "delete", Group: "agents.x-k8s.io", Resource: "sandboxes", Namespaced: true},
	{Name: "perm-list-sandboxclaims", Verb: "list", Group: extensionsAPIGroup, Resource: "sandboxclaims", Namespaced: true},
	{Name: "perm-create-sandboxclaims", Verb: "create", Group: extensionsAPIGroup, Resource: "sandboxclaims", Namespaced: true},
	{Name: "perm-delete-sandboxclaims", Verb: "delete", Group: extensionsAPIGroup, Resource: "sandboxclaims", Namespaced: true},
	{Name: "perm-get-sandboxtemplates", Verb: "get", Group: extensionsAPIGroup, Resource: "sandboxtemplates", Namespaced: true},
	{Name: "perm-list-sandboxwarmpools", Verb: "list", Group: extensionsAPIGroup, Resource: "sandboxwarmpools", Namespaced: true},
	{Name: "perm-list-pods", Verb: "list", Resource: "pods", Namespaced: true},
	{Name: "perm-create-pods-exec", Verb: "create", Resource: "pods", Subresource: "exec", Namespaced: true},
	{Name: "perm-get-persistentvolumeclaims", Verb: "get", Resource: "persistentvolumeclaims", Namespaced: true},
	{Name: "perm-patch-persistentvolumeclaims", Verb: "patch", Resource: "persistentvolumeclaims", Namespaced: true},
	{Name: "perm-list-resourcequotas", Verb: "list", Resource: "resourcequotas", Namespaced: true},
	{Name: "perm-list-limitranges", Verb: "list", Resource: "limitranges", Namespaced: true},
	{Name: "perm-list-storageclasses", Verb: "list", Group: "storage.k8s.io", Resource: "storageclasses"},
	{Name: "perm-get-runtimeclasses", Verb: "get", Group: "node.k8s.io", Resource: "runtimeclasses"},
	{Name: "perm-get-namespaces", Verb: "get", Resource: "namespaces"},
}

func runKubectl(ctx context.Context, options globalOptions, args ...string) (string, error) {
	env := []string{}
	if options.Kubeconfig != "" {
		env = append(env, "KUBECONFIG="+options.Kubeconfig)
	}
	return runCaptured(ctx, options.Kubectl, args, env, nil)
}

func baseArgs(target KubeTarget, requestTimeout string, args ...string) []string {
	return append([]string{"--context", target.Context, "--request-timeout", requestTimeout}, args...)
}

func namespacedArgs(target KubeTarget, requestTimeout string, args ...string) []string {
	return append([]string{"--context", target.Context, "-n", target.Namespace, "--request-timeout", requestTimeout}, args...)
}

func okCheck(name, message string) CheckResult {
	return CheckResult{Name: name, OK: true, Message: message}
}

func failedCheck(name, message string) CheckResult {
	return CheckResult{Name: name, Message: message}
}

func (a *App) runCheck(ctx context.Context, target KubeTarget, options globalOptions, requestTimeout time.Duration) CheckReport {
	requestTimeoutArg := requestTimeout.String()
	perCall := requestTimeout + 5*time.Second
	perCall = max(perCall, 15*time.Second)
	attempt := func(name, action string, args []string) (string, *CheckResult) {
		callCtx, cancel := context.WithTimeout(ctx, perCall)
		defer cancel()
		out, err := runKubectl(callCtx, options, args...)
		if err == nil {
			return out, nil
		}
		result := failedCheck(name, "could not "+action)
		return out, &result
	}

	checks := []CheckResult{}
	versions, connectivity := attempt(
		"api-connectivity",
		"reach the Kubernetes API with the saved context",
		baseArgs(target, requestTimeoutArg, "api-versions"),
	)
	served := sets.New[string]()
	if connectivity != nil {
		checks = append(checks, *connectivity)
	} else {
		checks = append(checks, okCheck("api-connectivity", "API reachable with saved context"))
		served.Insert(strings.Fields(versions)...)
	}
	probe := func(name, action string, args []string, checkOutput func(string) CheckResult) CheckResult {
		out, failure := attempt(name, action, args)
		if failure != nil {
			return *failure
		}
		return checkOutput(out)
	}
	online := connectivity == nil
	for _, expectation := range expectedAPIGroups {
		if online && !served.Has(expectation.Group+"/v1beta1") {
			checks = append(checks, failedCheck(expectation.Check, expectation.Label+" is not served"))
			continue
		}
		checks = append(checks, probe(
			expectation.Check,
			"discover "+expectation.Label,
			baseArgs(target, requestTimeoutArg, "api-resources", "--api-group="+expectation.Group, "-o", "name"),
			expectation.check,
		))
	}

	checks = append(checks,
		probe(runtimeClassCheckName, "read RuntimeClass gvisor", baseArgs(target, requestTimeoutArg, "get", "runtimeclass", "gvisor", "-o", "json"), runtimeClassCheck),
		probe("storage", "list StorageClasses", baseArgs(target, requestTimeoutArg, "get", "storageclass", "-o", "json"), storageCheck),
		probe("namespace", fmt.Sprintf("read namespace %q", target.Namespace), baseArgs(target, requestTimeoutArg, "get", "namespace", target.Namespace, "-o", "json"), func(string) CheckResult {
			return okCheck("namespace", fmt.Sprintf("namespace %q exists", target.Namespace))
		}),
		probe(quotaCheckName, "read ResourceQuotas in target namespace", namespacedArgs(target, requestTimeoutArg, "get", "resourcequota", "-o", "json"), quotaCheck),
	)
	out, failure := attempt(limitsCheckName, "read LimitRanges in target namespace", namespacedArgs(target, requestTimeoutArg, "get", "limitrange", "-o", "json"))
	limits := limitsCheck(out)
	if failure != nil {
		limits = *failure
		limits.Advisory = true
	}
	checks = append(checks, limits)
	checks = append(checks, checkAccess(ctx, target, options, requestTimeoutArg, perCall)...)
	report := CheckReport{Context: target.Context, Namespace: target.Namespace, OK: true, Checks: checks, CheckedAt: a.now().UTC()}
	for _, check := range checks {
		if !check.OK && !check.Advisory {
			report.OK = false
		}
	}
	return report
}

var (
	computeQuotaKeys = sets.New("pods", "cpu", "memory", "requests.cpu", "requests.memory", "limits.cpu", "limits.memory", "requests.ephemeral-storage", "limits.ephemeral-storage")
	storageQuotaKeys = sets.New("persistentvolumeclaims", "requests.storage")
)

type apiGroupExpectation struct {
	Check     string
	Group     string
	Label     string
	Resources []string
}

var expectedAPIGroups = []apiGroupExpectation{
	{
		Check:     "agents-api",
		Group:     "agents.x-k8s.io",
		Label:     "Sandbox API agents.x-k8s.io/v1beta1",
		Resources: []string{"sandboxes"},
	},
	{
		Check: "extensions-api",
		Group: extensionsAPIGroup,
		Label: "Sandbox extensions API extensions.agents.x-k8s.io/v1beta1",
		Resources: []string{
			"sandboxclaims",
			"sandboxtemplates",
			"sandboxwarmpools",
		},
	},
}

func (expectation apiGroupExpectation) check(out string) CheckResult {
	missing := []string{}
	lowered := strings.ToLower(out)
	for _, resource := range expectation.Resources {
		if !strings.Contains(lowered, resource) {
			missing = append(missing, resource)
		}
	}
	if len(missing) != 0 {
		return failedCheck(expectation.Check, expectation.Label+" served but missing: "+strings.Join(missing, ", "))
	}
	return okCheck(expectation.Check, expectation.Label+" served")
}

func runtimeClassCheck(out string) CheckResult {
	var value struct {
		Handler string `json:"handler"`
	}
	if json.Unmarshal([]byte(out), &value) != nil {
		return failedCheck(runtimeClassCheckName, "RuntimeClass gvisor returned unreadable JSON")
	}
	if !strings.Contains(strings.ToLower(value.Handler), "runsc") {
		return failedCheck(runtimeClassCheckName, fmt.Sprintf("RuntimeClass gvisor handler is %q, want runsc", value.Handler))
	}
	return okCheck(runtimeClassCheckName, fmt.Sprintf("RuntimeClass gvisor present (handler %s)", value.Handler))
}

type storageClass struct {
	Metadata struct {
		Name        string            `json:"name"`
		Annotations map[string]string `json:"annotations"`
	} `json:"metadata"`
}

func storageCheck(out string) CheckResult {
	var value struct {
		Items []storageClass `json:"items"`
	}
	if json.Unmarshal([]byte(out), &value) != nil {
		return failedCheck("storage", "StorageClass list returned unreadable JSON")
	}
	if len(value.Items) == 0 {
		return failedCheck("storage", "no StorageClasses available for sandbox homes")
	}
	names := make([]string, len(value.Items))
	label := ""
	for i, item := range value.Items {
		names[i] = item.Metadata.Name
		if label == "" && item.Metadata.Annotations["storageclass.kubernetes.io/is-default-class"] == "true" {
			label = " (default " + item.Metadata.Name + ")"
		}
	}
	return okCheck("storage", fmt.Sprintf("%d StorageClass(es): %s%s", len(names), strings.Join(names, ", "), label))
}

func checkAccess(ctx context.Context, target KubeTarget, options globalOptions, req string, perCall time.Duration) []CheckResult {
	permissions := make([]CheckResult, len(requiredAccess))
	group := new(errgroup.Group)
	group.SetLimit(6)
	for i, access := range requiredAccess {
		group.Go(func() error {
			permissions[i] = permissionCheck(ctx, target, options, req, perCall, access)
			return nil
		})
	}
	_ = group.Wait()
	return permissions
}

type resourceQuota struct {
	Metadata struct {
		Name string `json:"name"`
	} `json:"metadata"`
	Spec struct {
		Hard map[string]string `json:"hard"`
	} `json:"spec"`
}

func quotaCheck(raw string) CheckResult {
	var value struct {
		Items []resourceQuota `json:"items"`
	}
	if json.Unmarshal([]byte(raw), &value) != nil {
		return failedCheck(quotaCheckName, "ResourceQuota list returned unreadable JSON")
	}
	if len(value.Items) == 0 {
		return failedCheck(quotaCheckName, "no ResourceQuota in target namespace; set explicit compute/storage budgets via the Helm chart")
	}
	names := make([]string, len(value.Items))
	seen := sets.New[string]()
	for i, item := range value.Items {
		names[i] = item.Metadata.Name
		for key := range item.Spec.Hard {
			seen.Insert(key)
		}
	}
	hasCompute := seen.Intersection(computeQuotaKeys).Len() != 0
	hasStorage := seen.Intersection(storageQuotaKeys).Len() != 0
	if hasCompute && hasStorage {
		return okCheck(quotaCheckName, "ResourceQuota present: "+strings.Join(names, ", "))
	}
	missing := []string{}
	if !hasCompute {
		missing = append(missing, "compute (pods, cpu, memory)")
	}
	if !hasStorage {
		missing = append(missing, "storage")
	}
	hard := strings.Join(slices.Sorted(maps.Keys(seen)), ", ")
	if hard == "" {
		hard = "empty"
	}
	return failedCheck(quotaCheckName, fmt.Sprintf("ResourceQuota %s sets no %s budget (hard: %s)", strings.Join(names, ", "), strings.Join(missing, " and "), hard))
}

func limitsCheck(raw string) CheckResult {
	var value struct {
		Items []jsontext.Value `json:"items"`
	}
	if json.Unmarshal([]byte(raw), &value) != nil {
		result := failedCheck(limitsCheckName, "LimitRange list returned unreadable JSON")
		result.Advisory = true
		return result
	}
	if len(value.Items) == 0 {
		result := failedCheck(limitsCheckName, "no LimitRange in target namespace")
		result.Advisory = true
		return result
	}
	return okCheck(limitsCheckName, fmt.Sprintf("%d LimitRange(s) present", len(value.Items)))
}

func permissionCheck(
	ctx context.Context,
	target KubeTarget,
	options globalOptions,
	requestTimeout string,
	timeout time.Duration,
	access accessCheck,
) CheckResult {
	shown := access.qualified()
	args := []string{"auth", "can-i", access.Verb, access.resourceArg()}
	if access.Subresource != "" {
		args = append(args, "--subresource="+access.Subresource)
	}
	if access.Namespaced {
		args = namespacedArgs(target, requestTimeout, args...)
	} else {
		args = baseArgs(target, requestTimeout, args...)
	}
	callCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	out, err := runKubectl(callCtx, options, args...)
	if command, ok := errors.AsType[*commandError](err); ok && strings.TrimSpace(command.Stdout) != "" {
		out = command.Stdout
	}
	answer := strings.ToLower(strings.TrimSpace(out))
	if answer == "yes" {
		return okCheck(access.Name, fmt.Sprintf("can %s %s", access.Verb, shown))
	}
	if answer == "no" || strings.HasPrefix(answer, "no ") {
		scope := "cluster scope"
		if access.Namespaced {
			scope = "namespace"
		}
		result := failedCheck(access.Name, fmt.Sprintf("cannot %s %s in %s", access.Verb, shown, scope))
		result.Advisory = access.Advisory
		return result
	}
	message := fmt.Sprintf("could not check permission %s %s", access.Verb, shown)
	if err == nil {
		message += ": kubectl returned an unexpected response"
	}
	result := failedCheck(access.Name, message)
	result.Advisory = access.Advisory
	return result
}
